osu.Desktop/OsuGameDesktop.cs

```diff
@@ -102,35 +100,21 @@ protected override UpdateManager CreateUpdateManager()
             if (!string.IsNullOrEmpty(packageManaged))
                 return new NoActionUpdateManager();
 
-            switch (RuntimeInfo.OS)
-            {
-                case RuntimeInfo.Platform.Windows:
-                    Debug.Assert(OperatingSystem.IsWindows());
-
-                    return new SquirrelUpdateManager();
-
-                default:
-                    return new SimpleUpdateManager();
-            }
+            return new VelopackUpdateManager();
         }
 
         public override bool RestartAppWhenExited()
         {
-            switch (RuntimeInfo.OS)
+            try
             {
-                case RuntimeInfo.Platform.Windows:
-                    Debug.Assert(OperatingSystem.IsWindows());
-
-                    // Of note, this is an async method in squirrel that adds an arbitrary delay before returning
-                    // likely to ensure the external process is in a good state.
-                    //
-                    // We're not waiting on that here, but the outro playing before the actual exit should be enough
-                    // to cover this.
-                    Squirrel.UpdateManager.RestartAppWhenExited().FireAndForget();
-                    return true;
+                Velopack.UpdateExe.Start();
```

Comment:
This call is now blocking and will cause a stutter in the game.

We can avoid this with an assumption that the outro sequence will take longer than the updater process takes to start up:

```diff
diff --git a/osu.Desktop/OsuGameDesktop.cs b/osu.Desktop/OsuGameDesktop.cs
index 94f6ef0fc3..c75a3f0a1a 100644
--- a/osu.Desktop/OsuGameDesktop.cs
+++ b/osu.Desktop/OsuGameDesktop.cs
@@ -5,6 +5,7 @@
 using System.IO;
 using System.Reflection;
 using System.Runtime.Versioning;
+using System.Threading.Tasks;
 using Microsoft.Win32;
 using osu.Desktop.Performance;
 using osu.Desktop.Security;
@@ -18,6 +19,7 @@
 using osu.Framework.Allocation;
 using osu.Game.IO;
 using osu.Game.IPC;
+using osu.Game.Online.Multiplayer;
 using osu.Game.Performance;
 using osu.Game.Utils;
 
@@ -105,16 +107,8 @@ protected override UpdateManager CreateUpdateManager()
 
         public override bool RestartAppWhenExited()
         {
-            try
-            {
-                Velopack.UpdateExe.Start();
-                return true;
-            }
-            catch (Exception e)
-            {
-                Logger.Error(e, "Failed to restart application");
-                return base.RestartAppWhenExited();
-            }
+            Task.Run(() => Velopack.UpdateExe.Start()).FireAndForget();
+            return true;
         }
 
         protected override void LoadComplete()

```

Or if we actually need to wait for the process... things are going to get quite a bit more complicated.

@smoogipoo thoughts on this? I think I'd just propose the above to match what we were already doing.

Reply:
The above solution works fine. In my [discussion](https://discord.com/channels/188630481301012481/188630652340404224/1255899633687859270) with caesay, he implemented a restart implementation into velopack which should wait 30 seconds for the assembly to exit on its own, otherwise it will kill it. We don't necessarily need to assume anything here, the updater process will wait for us.

Though, I'm a little iffy about doing `FireAndForget`. I would still like to handle exceptions in starting the restarter. Best not to leave the caller of the function in a state of confusion as to whether the game will actually restart or not.

Reply:
I think it's fine to match what we were already doing.

Reply:
> No need to worry about pacing/race conditions here.

The fail case would the update process *not being started* in time, not the game *not being killed in time*. It's a different concern, but as I mentioned it's very unlikely to cause a problem. Worst case the user has to start osu! again because it didn't do so automatically.

Reply:
There is nothing blocking about UpdateExe.Start. there does appear to be a Thread.Sleep(300) in UpdateExe.Apply but I could refactor this to a Task.Delay

Reply:
Hmm, I have noticed a pause on macOS. I've have to profile to be sure, but I think `Process.Start` can block for short time.

Reply:
@caesay profiled this. as you say, it's mostly the sleep:

![JetBrains Rider 2024-09-04 at 06 50 17](https://github.com/user-attachments/assets/f9d84e63-bc35-4ec4-bc27-951a9a7551f2)

Of note, the overhead of starting the process *may still cause a frame stutter for users with higher refresh rates*. But the sleep part is the bigger issue.

Making this flow async would probably be the best outcome.